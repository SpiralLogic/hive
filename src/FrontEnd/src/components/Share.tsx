import '../css/share.css';
import { FunctionComponent, h } from 'preact';
import Modal from './Modal';

type Props = { url: string; setShowShare: (value: boolean) => void };

const fallbackCopyTextToClipboard = (text: string) => {
  const currentFocus = document.activeElement as HTMLElement;
  const textArea = document.createElement('textarea');
  textArea.value = text;

  textArea.style.top = '0';
  textArea.style.left = '0';
  textArea.style.position = 'fixed';

  document.body.appendChild(textArea);
  textArea.focus();
  textArea.select();

  try {
    document.execCommand('copy');
    if (currentFocus && currentFocus.focus) currentFocus.focus();
    return true;
  } catch (e) {
    //  if (process.env.NODE_ENV !== 'production') console.error(e);
  } finally {
    document.body.removeChild(textArea);
  }
  return false;
};

const Share: FunctionComponent<Props> = (props) => {
  const opponentGame = {
    title: 'Hive board game',
    text: 'Join me in a game of the Hive!',
    url: props.url,
  };
  try {
    navigator.share(opponentGame).then();
    return null;
  } catch {
    if (navigator.clipboard && navigator.clipboard.writeText) {
      navigator.clipboard.writeText(opponentGame.url).then();
    } else {
      if (!fallbackCopyTextToClipboard(opponentGame.url)) return null;
    }
  }
  return (
    <Modal name="share" onClose={() => props.setShowShare(false)}>
      <p>Opponent's link has been copied to clipboard!</p>
      <button title="Close" onClick={() => props.setShowShare(false)}>
        Close
      </button>{' '}
    </Modal>
  );
};
Share.displayName = 'Share';
export default Share;
