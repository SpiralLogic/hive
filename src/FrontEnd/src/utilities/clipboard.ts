const fallbackCopyTextToClipboard = (text: string) => {
  const currentFocus = document.activeElement as HTMLElement;
  const textArea = document.createElement('textarea');
  textArea.value = text;

  textArea.style.top = '0';
  textArea.style.left = '0';
  textArea.style.position = 'fixed';

  document.body.appendChild(textArea);
  textArea.focus();
  textArea.select();
  try {
    document.execCommand('copy');
    if (currentFocus && currentFocus.focus) currentFocus.focus();
    return true;
  } catch (e) {
    //  if (process.env.NODE_ENV !== 'production') console.error(e);
  } finally {
    document.body.removeChild(textArea);
  }
  return false;
};

export const shareGame = async (url: string) => {
  const opponentGame = {
    title: 'Hive board game',
    text: 'Join me in a game of the Hive!',
    url,
  };
  try {
    await navigator.share(opponentGame);
    return false;
  } catch {
    if (navigator.clipboard && navigator.clipboard.writeText) {
      try {
        await navigator.clipboard.writeText(opponentGame.url);
        return true;
      } catch {
        return fallbackCopyTextToClipboard(opponentGame.url);
      }
    }
  }
  return false;
};
