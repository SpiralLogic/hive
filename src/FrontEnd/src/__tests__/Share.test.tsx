import { h } from 'preact';
import { mockClipboard, mockExecCommand, mockShare } from './helpers/clipboard';
import { render, screen } from '@testing-library/preact';
import { renderElement } from './helpers';
import Share from '../components/Share';

describe('Share tests', () => {
  test('if available share API is called', () => {
    const restore = mockShare();
    render(<Share setShowShare={() => ({})} />);
    expect(navigator.share).toBeCalledWith(expect.objectContaining({ url: expect.stringMatching(/.*1$/) }));
    restore();
  });

  test('Click copies opponent link to clipboard with navigator', () => {
    const restore = mockClipboard();
    render(<Share setShowShare={() => ({})} />);
    expect(navigator.clipboard.writeText).toBeCalledWith(expect.stringMatching(/.*1$/));
    restore();
  });

  test('Click copies opponent link to clipboard with exec command', () => {
    const restore = mockExecCommand();
    render(<Share setShowShare={() => ({})} />);
    expect(document.execCommand).toBeCalledWith('copy');
    restore();
  });

  describe('Share snapshot tests', () => {
    test('snapshot', () => {
      expect(renderElement(<Share setShowShare={() => ({})} />)).toMatchSnapshot();
    });
  });
});
