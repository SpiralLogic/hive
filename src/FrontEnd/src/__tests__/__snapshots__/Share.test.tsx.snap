// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Share tests Share snapshot tests snapshot 1`] = `
<div
  class="modal"
>
  <div
    class="share"
    role="dialog"
    title="share"
  >
    <p>
      Opponent's link has been copied to clipboard!
    </p>
    <button
      title="Close"
    >
      Close
    </button>
  </div>
</div>
`;
